crates/polars-plan/src/plans/conversion/join.rs

```diff
@@ -467,137 +464,129 @@ fn resolve_join_where(
-                }
-
-                // Fetch them again in case they were rewritten.
-                let left = expr_arena.get(left_node).clone();
-                let right = expr_arena.get(right_node).clone();
-
-                let resolve_dtype = |ae: &AExpr, node: Node| -> PolarsResult<DataType> {
-                    ae.to_dtype(schema_merged, Context::Default, expr_arena)
-                        .map_err(|e| {
-                            e.context(
-                                format!(
-                                    "could not resolve dtype of join_where predicate (expr: {})",
-                                    node_to_expr(node, expr_arena),
-                                )
-                                .into(),
-                            )
-                        })
-                };
-
-                let dtype_left = resolve_dtype(&left, left_node)?;
-                let dtype_right = resolve_dtype(&right, right_node)?;
-
-                // Note: We only upcast the sides if the expr output dtype is Boolean (i.e. `op` is
-                // a comparison), otherwise the output may change.
-
-                if let Some(dtype) =
-                    get_numeric_upcast_supertype_lossless(&dtype_left, &dtype_right)
-                        .filter(|_| op.is_comparison())
-                {
-                    // We have unique references to these nodes (they are created by this function),
-                    // so we can mutate in-place without causing side effects somewhere else.
-                    let expr = expr_arena.add(expr_arena.get(left_node).clone());
-                    expr_arena.replace(
-                        left_node,
-                        AExpr::Cast {
-                            expr,
-                            dtype: dtype.clone(),
-                            options: CastOptions::Overflowing,
-                        },
-                    );
-
-                    let expr = expr_arena.add(expr_arena.get(right_node).clone());
-                    expr_arena.replace(
-                        right_node,
-                        AExpr::Cast {
-                            expr,
-                            dtype,
-                            options: CastOptions::Overflowing,
-                        },
-                    );
-                } else {
-                    polars_ensure!(
-                        dtype_left == dtype_right,
-                        SchemaMismatch:
-                        "datatypes of join_where comparison don't match - {} on left does not match {} on right \
-                        (expr: {})",
-                        dtype_left, dtype_right, node_to_expr(ae_node, expr_arena),
-                    )
-                }
-            },
-            ae => ae.inputs_rev(stack),
-        }
+    // Ensure that all binary comparisons that use both tables are lossless.
+    let mut to_replace = Vec::<(Node, DataType)>::with_capacity(expr_arena.len());
```

Comment:
This seems like a gross overallocation? The exp_arena can be very long. I also think we can amortize the allocation and pass a `&mut Vec` that is reused for every `e` in `predicates.`.